atl-checker/src/algorithms/certain_zero/search_strategy/linear_optimize.rs

```diff
@@ -0,0 +1,515 @@
+use crate::algorithms::certain_zero::search_strategy::{SearchStrategyBuilder, SearchStrategy};
+use minilp::{Problem, OptimizationDirection, ComparisonOp};
+use priority_queue::PriorityQueue;
+use std::collections::HashMap;
+use BinaryOpKind::{Addition,
+                   Multiplication,
+                   Subtraction,
+                   Division,
+                   Equality,
+                   Inequality,
+                   GreaterThan,
+                   LessThan,
+                   GreaterOrEqual,
+                   LessOrEqual,
+                   And,
+                   Or,
+                   Xor,
+                   Implication, };
+
+/// Holds extracted linear expressions from the formula in ATLVertex
+#[derive(Eq, PartialEq, Hash, Clone)]
+struct LinearExpression {
+    pub symbol: SymbolIdentifier,
+    pub constant: i32,
+    pub operation: BinaryOpKind,
+}
+
+/// Search strategy using ideas from linear programming to order next vertices,
+/// based on distance from the vertex to a region that borders the line between true/false in the formula.
+/// Has a cache to hold previously computed results, and uses a priority queue with
+/// distances as priority (lowest distance has highest priority)
+pub struct LinearOptimizeSearch {
+    queue: PriorityQueue<Edge<ATLVertex>, i32>,
+    game: IntermediateLCGS,
+    cache: HashMap<(Vec<LinearExpression>, usize), i32>,
+}
+
+impl LinearOptimizeSearch {
+    pub fn new(game: IntermediateLCGS) -> LinearOptimizeSearch {
+        LinearOptimizeSearch {
+            queue: PriorityQueue::new(),
+            game,
+            cache: HashMap::new(),
+        }
+    }
+}
+
+/// A SearchStrategyBuilder for building the LinearOptimizeSearch strategy.
+pub struct LinearOptimizeSearchBuilder {
+    pub game: IntermediateLCGS,
+}
+
+impl SearchStrategyBuilder<ATLVertex, LinearOptimizeSearch> for LinearOptimizeSearchBuilder {
+    fn build(&self) -> LinearOptimizeSearch {
+        LinearOptimizeSearch::new(self.game.clone())
+    }
+}
+
+impl SearchStrategy<ATLVertex> for LinearOptimizeSearch {
+    /// Simply returns the edge with highest priority (i.e lowest distance)
+    fn next(&mut self) -> Option<Edge<ATLVertex>> {
+        let edge = self.queue.pop();
+        if edge.is_some() {
+            Some(edge.unwrap().0)
+        } else { None }
+    }
+
+    /// Takes a Vec of Edges holding ATLVertices, and puts these in the queue,
+    /// based on the calculated distance from its state to acceptance region from formula
+    fn queue_new_edges(&mut self, edges: Vec<Edge<ATLVertex>>) {
+        for edge in edges {
+            let distance = self.get_distance_in_edge(&edge);
+
+            // Add edge and distance to queue
+            if let Some(dist) = distance {
+                self.queue.push(edge, -dist as i32);
+            } else {
+                // Todo what should default value be, if cannot be calculated?
+                self.queue.push(edge, 0);
```

Comment:
Yeah, this is a big question/problem because it affects the performance by a lot, but there isn't a correct answer. This is one of the reasons I think we should sort the targets instead of prioritising (incomparable) edges.